### ISSUE TITLE
Got wrong result when mapping Jpa Entity to DTO 

### ISSUE THREAD

**Issue by landcom** [modelmapper/modelmapper#586]
hi
I got a strange problem when I try to convert a Jpa Entity to DTO by modelmapper.map(). When some members of the source class are the same Type  , and the source class also have a collection of this kind of Type. These members in mapped result may be overridden by a same value. My issue is similar to #581, but it only happens when a Jpa entity is converted to DTO by modelMapper. I wonder if this is a bug or if I've mixed up something.

Comment of @lolo101 is more clear description about this issue.

For example:
I have two entities, User and Summary:
```java
@Entity
@Getter
@Setter
@ToString
@Table(name = "test_user")
public class User {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Integer id;
    @Column(name = "username")
    String name;
}

@Entity
@Getter
@Setter
@ToString
@Table(name = "test_summary")
public class Summary {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Integer id;

    @OneToMany
    List<User> readers;

    @ManyToOne
    User summarizer;

    @ManyToOne
    User signer;
}

```
two DTO, UserDTO and SummaryDTO. 
The SummaryDTO include UserDTO and collection of UserDTO. 

```java
@Data
public class UserDTO {
    Integer id;
    String name;
}

@Data
public class SummaryDTO {
    List<UserDTO> readers;
    UserDTO summarizer;
    UserDTO signer;
}
```

I have two User, Bob and Alice, and I add a Summary record like this in my database.
id =1
readers = [User(id=1, name=Bob), User(id=2, name=Alice)]
summarizer = User(id=1, name=Bob)
signer = User(id=2, name=Alice)

Then I ran this to test the result:

```java
@Transactional
    public void testModelMapper(){
        Summary summary = summaryRepo.findById(1).orElse(null);
        if(summary != null) {
            System.out.println(summary.toString());
            System.out.println(modelMapper.map(summary, SummaryDTO.class).toString());
        }
    }
```
The output is：
Summary(id=1, readers=[User(id=1, name=Bob), User(id=2, name=Alice)], summarizer=User(id=2, name=Alice), signer=User(id=1, name=Bob))
SummaryDTO(readers=[UserDTO(id=2, name=Alice), UserDTO(id=2, name=Alice)], summarizer=UserDTO(id=2, name=Alice), signer=UserDTO(id=2, name=Alice))

 The entity's data is exactly correct，but the User data in the mapped DTO is all overwritten by Alice. 

PS:
This is my ModelMapper config:

```java 
@Configuration
public class ModelMapperConfig{
    @Bean
    public ModelMapper modelMapper() {
        ModelMapper modelMapper = new ModelMapper();
        modelMapper.getConfiguration().setMatchingStrategy(MatchingStrategies.STRICT).setPropertyCondition(Conditions.isNotNull());
        return modelMapper;
    }
}
```

Add this to config will fix this problem, but it seems redundant, I don't know how it works.
```java
modelMapper.createTypeMap(Summary.class, SummaryDTO.class).addMappings(mapping -> mapping.map(Summary::getSummarizer, SummaryDTO::setSummarizer));
```

**Comment by lolo101** [comment]
I believe I met the same issue.
I built a small demo of the bug here : https://github.com/lolo101/modelmapper-bug

The bug occurs in a very specific case :
- you need to have a collection of some Type,
- another property of the same Type that is equal to some element of the collection
- another property of the same Type that is different from the previous property

**Comment by KagariSan** [comment]
@lolo101 As far as I remember I meet some of this bug in my production code which fixed by using specific 
```modelmapper.typeMap(Source.class, Target.class).map(source);```
Don't know what the cause then...

**Comment by chhsiao90** [comment]
Duplicates to #581, should be fixed in #593

**Comment by chhsiao90** [comment]
ModelMapper v2.4.0 released.

I suggest to disable [Prefer Nested Properties](http://modelmapper.org/javadoc/org/modelmapper/config/Configuration.html#setPreferNestedProperties-boolean-) for mapping models with circular reference.

You can apply this option to single type map.
```
modelMapper.createTypeMap(SourceTree.class, DestinationTree.class,
    modelMapper.getConfiguration().copy().setPreferNestedProperties(false));
```
